### ISSUE TITLE
Initializing both the models and bin data file retrains the model, instead of continuing training. 

### ISSUE THREAD

**Issue by nd7141** [microsoft/LightGBM#4786]
## Description
I want to run distributed CLI training and I want to run the training twice: the second time initializing the model from the first run. 

So what I do is I use parameter `init_model` the second time. 

What I found is that depending on whether I use another parameter `is_save_binary_file` the results are different. In particular, **if I use `is_save_binary_file=False`, then I get the continual training, that I expect.** 

Here I have loss curves for two runs, 100 iterations each. So you can see the line is smooth, meaning that the second run picked up the first model. 

![image](https://user-images.githubusercontent.com/4607709/140934227-4d3a1e70-9492-4449-af7e-5ca653f73c4a.png)

But, **if I use `is_save_binary_file=True`, then the second run retrains the model.**

![image](https://user-images.githubusercontent.com/4607709/140934418-172b1a5b-7b77-4f7a-83e0-a354a627221a.png)

I would expect that binary files should not affect initialization of the model. In particular, I would expect `is_save_binary_file=True` just creates convenient bin mappers for the future trees, that are independent of each run I use them for.

Is it a bug or expected behavior? How would I go about using both parameters `init_model` and `is_save_binary_file`? 

It would be useful to use binary files, as they are very quick to read, and being able to initialize the model as well. Is it difficult to do? Can you give some pointers where this code logic is? 

## Reproducible example
This example gives the setting for distributed CLI version. Just try to train the second run with and without bin files. 
In particular, 

Setting A. `is_save_binary_file=False`

Setting B. `is_save_binary_file=True`

In both settings, you need to have 2 runs, the first one that saves the model `output_model` and the second one that loads the model `input_model`.

## Environment info

LightGBM version or commit hash: 3.2.1

```shell
lightgbm config=train.conf
```

**Comment by jameslamb** [comment]
Thanks for the excellent write-up and for using LightGBM! I'll look into this some time in the next few days if other maintainers don't get to it before me.

**Comment by jmoralez** [comment]
This seems to be related to using the binary files themselves to train. When you use `is_save_binary_file = true` the first time it creates two files: `binary.train.bin` and `binary.test.bin`. The second time you train it realizes those files exist and reads from them instead and that's when the training starts from scratch. Moving these files somewhere else actually continues training. Not sure if this is expected though, maybe @shiyu1994 knows.

**Comment by nd7141** [comment]
@jmoralez Yes, I agree, but moving these bin files somewhere discards the whole point of generating them in the first place. The reason to use `is_save_binary_file = true` is to generate bin files in the 1st run so that it does not do expensive preprocessing during the 2nd training phase.

**Comment by jmoralez** [comment]
@nd7141 sorry I think I wasn't clear. What I meant was that the problem seems to be related with trying to continue training using a binary file as input, not with the argument `is_save_binary_file`. For example, doing `./lightgbm task=save_binary data=binary.train` generates the `binary.train.bin` file. If you train a model using this file it works fine, and if you then try to continue training using the trained model as `input_model` the training starts from scratch. I agree it would be very useful to use a binary file to continue training and I'd expect it to work.

**Comment by nd7141** [comment]
@jmoralez Yes, with that I agree. I expect that now the model somehow decides to start training anew if it loads a bin file instead of the original one, and this is wrong.

**Comment by shiyu1994** [comment]
@nd7141 Thanks for using LightGBM. And thank @jameslamb and @jmoralez for looking into this issue.
Actually, LightGBM will simply ignore the `input_model` when loading data from binary file. In this method which loads binary file, there's nothing done to handle `input_model`,
https://github.com/microsoft/LightGBM/blob/99e0a4bd7b9e7c557e593ff9172799822abc4b7d/src/io/dataset_loader.cpp#L316-L318
In comparison, when loading a data from raw text file, it will add the initial prediction values with the input model, see e.g.,
https://github.com/microsoft/LightGBM/blob/99e0a4bd7b9e7c557e593ff9172799822abc4b7d/src/io/dataset_loader.cpp#L1213-L1224

Ignoring `input_model` when loading from binary data has reasonable justification:
1. A `model` means something that can serve with new and raw data, so it contain no internal information of the training algorithm. For example, how to discretize the raw features into bins, how to group the features etc.
2. But a `binary` data means it is an internal representation of a raw dataset in LightGBM. So it is a format preprocessed by LightGBM, leaving out the raw feature information. It contains only discretized and grouped features, instead of the original one.
So, we can see that a model is incompatible with a binary data input inherently. Using initial prediction score should be an alternative in this case. But I'm not sure whether there's a way to specify initial score in CLI version for now. If not, we can add it as a feature request, since it is feasible to implement.

On the other hand, I think we do need a better documentation to inform the user that binary data input should not be used with `input_model`.

**Comment by shiyu1994** [comment]
> But I'm not sure whether there's a way to specify initial score in CLI version for now.

I just remind that, there IS a way to load initial score from CLI. You may try it.
https://lightgbm.readthedocs.io/en/latest/Parameters.html#continued-training-with-input-score

**Comment by shiyu1994** [comment]
Sorry for coming back so late. After checking the source code, I found that the `.init` file won't take any effect if the dataset is loaded from binary file. The files with `.init` suffix are only used when loading dataset from raw txt files. I've open a pull request to fix this. You may try with the branch now. Thanks for the feedback which helps to find an important feature request!

**Comment by shiyu1994** [comment]
The initial scores for binary data file should have suffix `.bin.init`.

**Comment by nd7141** [comment]
Sounds great! I will test it in the coming week. 

Do you know to use this branch, do I have to download the whole repo anew and then build it or there is a more incremental way to include these changes into my current build?

**Comment by shiyu1994** [comment]
You need to clone the branch, or at least merge the branch into your local branch. And then rebuild LightGBM.

**Comment by jameslamb** [comment]
Sorry, this was locked accidentally. Just unlocked it. We'd still love help with this feature!
